blackjax/mcmc/integrators.py

```diff
@@ -38,11 +40,154 @@ class IntegratorState(NamedTuple):
+
+    Returns
+    -------
+    integrator
+        Integrator function.
+    """
+
+    def one_step(state: IntegratorState, step_size: float):
+        position, momentum, _, logdensity_grad = state
+        # auxiliary infomation generated during integration for diagnostics. It is updated
+        # by the momentum_update_fn and position_update_fn at each call
+        momentum_update_info = None
+        position_update_info = None
+        for i, coef in enumerate(coefficients[:-1]):
+            if i % 2 == 0:
+                momentum, kinetic_grad, momentum_update_info = momentum_update_fn(
+                    momentum,
+                    logdensity_grad,
+                    step_size,
+                    coef,
+                    momentum_update_info,
+                    is_last_call=False,
+                )
+            else:
+                (
+                    position,
+                    logdensity,
+                    logdensity_grad,
+                    position_update_info,
+                ) = position_update_fn(
+                    position,
+                    kinetic_grad,
+                    step_size,
+                    coef,
+                    position_update_info,
+                )
+        # Separate the last steps to short circuit the computation of the kinetic_grad
+        momentum, kinetic_grad, momentum_update_info = momentum_update_fn(
+            momentum,
+            logdensity_grad,
+            step_size,
+            coefficients[-1],
+            momentum_update_info,
+            is_last_call=True,
+        )
+        return format_output_fn(
+            position,
+            momentum,
+            logdensity,
+            logdensity_grad,
+            kinetic_grad,
+            position_update_info,
+            momentum_update_info,
+        )
+
+    return one_step
+
+
 def new_integrator_state(logdensity_fn, position, momentum):
     logdensity, logdensity_grad = jax.value_and_grad(logdensity_fn)(position)
     return IntegratorState(position, momentum, logdensity, logdensity_grad)
 
 
+def euclidean_position_update_fn(logdensity_fn: Callable):
+    logdensity_and_grad_fn = jax.value_and_grad(logdensity_fn)
+
+    def update(
+        position: ArrayTree,
+        kinetic_grad: ArrayTree,
+        step_size: float,
+        coef: float,
+        auxiliary_info=None,
+    ):
+        del auxiliary_info
+        new_position = jax.tree_util.tree_map(
+            lambda x, grad: x + step_size * coef * grad,
+            position,
+            kinetic_grad,
+        )
+        logdensity, logdensity_grad = logdensity_and_grad_fn(new_position)
+        return new_position, logdensity, logdensity_grad, None
+
+    return update
+
+
+def euclidean_momentum_update_fn(kinetic_energy_fn: EuclideanKineticEnergy):
+    kinetic_energy_grad_fn = jax.grad(kinetic_energy_fn)
+
+    def update(
+        momentum: ArrayTree,
+        logdensity_grad: ArrayTree,
+        step_size: float,
```

Comment:
Since step_size and coeff are always multiplied together, why not just treat them as a single step_size argument?

Reply:
I am thinking of if we are going to add `sigma` (inverse_mass_matrix) to MCLMC intergrators, we need some way to make sure the `step_size * coef` is general. Doing that operation in the function is more general as you can do e.g., tree_map(fn, step_size, ...) with coef inside the function.